In [1]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [2]:
# Load the given datasets
categorical_data = pd.read_csv('chronic_kidney_disease_categorical.csv')
numerical_data = pd.read_csv('chronic_kidney_disease_numerical.csv')
# Print the data
print(categorical_data.head())
print(numerical_data.head())

   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresent   no   no   

  cad appet   pe  ane  Target  
0  no  good  yes   no     ckd  
1  no  good   no   no  notckd  
2  no  good  yes   no     ckd  
3  no  good   no  yes     ckd  
4  no  good   no   no  notckd  
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     76411

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

## Numerical Features:
1. **Age**: This represents the individual's age in years, giving a straightforward numerical value.
2. **Blood Pressure (bp)**: This is the blood pressure measurement in mm/Hg, another numerical value.
3. **Blood Glucose Random (bgr)**: This indicates the random blood glucose level measured in mg/dl.
4. **Blood Urea (bu)**: This is the level of urea in the blood, measured in mg/dl.
5. **Serum Creatinine (sc)**: This pertains to the serum creatinine level, also in mg/dl.
6. **Sodium (sod)**: This is the sodium concentration measured in mEq/L.
7. **Potassium (pot)**: This is the potassium level, also measured in mEq/L.
8. **Hemoglobin (hemo)**: This level of hemoglobin in grams per deciliter (gms).
9. **Packed Cell Volume (pcv)**: This represents the volume percentage of red blood cells in blood.
10. **White Blood Cell Count (wbcc)**: This is the number of white blood cells per cubic millimeter (cells/cumm).
11. **Red Blood Cell Count (rbcc)**: This represents the number of red blood cells in millions per cubic millimeter (millions/cmm).

## Categorical Features:
1. **Specific Gravity (sg)**: This features nominal categories ranging from 1.005 to 1.025, indicating the concentration of substances in urine.
2. **Albumin (al)**: This indicates levels ranging from 0 to 5, describing the amount of albumin present in urine.
3. **Sugar (su)**: This represents levels from 0 to 5, indicating the amount of sugar in urine.
4. **Red Blood Cells (rbc)**: This feature has categories for normal and abnormal.
5. **Pus Cell (pc)**: This indicates whether pus cells are normal or abnormal.
6. **Pus Cell Clumps (pcc)**: This denotes whether pus cell clumps are present or not.
7. **Bacteria (ba)**: This indicates if bacteria are present or not.
8. **Hypertension (htn)**: This feature indicates whether the individual has hypertension (yes or no).
9. **Diabetes Mellitus (dm)**: This specifies if the individual has diabetes or not.
10. **Coronary Artery Disease (cad)**: This shows whether the individual has coronary artery disease (yes or no).
11. **Appetite (appet)**: This describes if the person's appetite is good or poor.
12. **Pedal Edema (pe)**: This indicates whether the individual experiences pedal edema (yes or no).
13. **Anemia (ane)**: This shows if the individual has anemia (yes or no).

## Label:
- **Class**: This is the target variable we want to predict, indicating whether a patient has chronic kidney disease (ckd) or not (notckd).

## Data Types:
The dataset consists of both **numerical features** (which can take on a range of values and have meaningful arithmetic operations performed on them) and **categorical features** (which represent categories or groups and cannot be operated on arithmetically in a meaningful way).


# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
is_duplicate_num = numerical_data.drop_duplicates(subset=['unique_id'], keep='first')
print(is_duplicate_num)
print(is_duplicate_num.shape)
print(numerical_data)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
is_duplicate_cat = categorical_data.drop_duplicates(subset=['unique_id'], keep='first')
print(is_duplicate_cat)
print(is_duplicate_cat.shape)
print(categorical_data)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

In [16]:
#print only unqiue_id column, to check if both datasets have same unique_id

print(numerical_data['unique_id'])
print(categorical_data['unique_id'])

0      203694
1      938027
2      421471
3      764115
4      240975
        ...  
415    522912
416    475200
417    168242
418    474407
419    926033
Name: unique_id, Length: 420, dtype: int64
0      224481
1      992643
2      308740
3      450314
4      881763
        ...  
425    769512
426    872034
427    643007
428    840392
429    856244
Name: unique_id, Length: 430, dtype: int64
False


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, all the unique ids are not present in both datasets. This is because the two datasets represent different aspects of the same problem. The first dataset contains the patient's medical test results and the second dataset contains the patient's symptoms. Therefore, some patients may have undergone medical tests but not reported any symptoms, and vice versa.

No, not all unique IDs are present in both datasets. The categorical dataset has 430 unique IDs, while the numerical dataset has only 420, indicating that some records are missing in the numerical dataset. This discrepancy suggests that 10 records in the categorical dataset do not have corresponding entries in the numerical dataset. When the two datasets are combined, the rows corresponding to missing `unique_id`s from the numerical dataset will show `NaN` values for all numerical columns. Similarly, if any IDs were missing in the categorical dataset, the respective rows would have `NaN` values in the categorical columns. This helps to identify and highlight the incomplete data entries between the two datasets.

In [5]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
unique_ids_num = set(is_duplicate_num['unique_id'])
unique_ids_cat = set(is_duplicate_cat['unique_id'])

common_ids = unique_ids_num.intersection(unique_ids_cat)

#print set difference using difference() method
print(unique_ids_num.difference(unique_ids_cat))
print(unique_ids_cat.difference(unique_ids_num))
#Print the combined dataset
combined_data = pd.merge(is_duplicate_num, is_duplicate_cat, on='unique_id', how = 'outer')
print(combined_data)

set()
{208034, 240430, 488947, 325238, 254775, 292312, 298297, 663387, 948317, 266335}
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   N

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [6]:
# Calculate the percentage of rows that contain atleast one missing value
missing_values = combined_data.isnull().sum(axis=1)
missing_rows = missing_values[missing_values > 0].count()
total_rows = combined_data.shape[0]
percentage = (missing_rows/total_rows)*100
# Print %
print(percentage)
# Drop these rows from the dataset
combined_data = combined_data.dropna()
# Print the Dataset
print(combined_data)

61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc        

## Part 3.4: Sort the dataset according to the Labels

In [7]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
combined_data = combined_data.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
print(combined_data)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [8]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
combined_data = pd.get_dummies(combined_data, columns=categorical_columns, dtype=int)
# Print the dataset
print(combined_data)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: Using a single column for binary variables (e.g., 0 for "No" and 1 for "Yes") is more efficient than using two separate columns, such as `cad_no` and `cad_yes`. It prevents redundancy and avoids multicollinearity issues, which can complicate model interpretation and affect performance. Thus, consolidating binary variables into a single column is typically a better approach for modeling.

For multi-class categorical variables with more than two values, a single binary column is not sufficient. Instead, methods like one-hot encoding, which creates a separate binary column for each category, are necessary to represent all unique values. This approach ensures the model can distinguish between categories and capture their individual impact on the target variable, leading to better predictive performance.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [9]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in numerical_columns:
    mean = combined_data[column].mean()
    std = combined_data[column].std()
    lower_bound = mean - 3*std
    upper_bound = mean + 3*std
    combined_data = combined_data[(combined_data[column] >= lower_bound) & (combined_data[column] <= upper_bound)]
# Print the dataset
print(combined_data)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [10]:
# Normalize the all Numerical Attributes in the dataset.
for column in numerical_columns:
    mean = combined_data[column].mean()
    maxVal = combined_data[column].max()
    minVal = combined_data[column].min()
    combined_data[column] = (combined_data[column] - minVal)/(maxVal - minVal)
# Print the dataset
print(combined_data)
print(combined_data.describe())
#.describe() function to get the summary statistics of the data
print(combined_data.describe())

     unique_id       age    bp       bgr        bu        sc       sod  \
2       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
5       397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
7       438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
8       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
9       475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
148     486397  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
149     203466  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
150     482105  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
151     228849  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
152     996076  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
2    0.793103  0.000000  0.032258  ...  

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The 'unique_id' column is not an appropriate feature for modeling and prediction because it is just a unique identifier for each patient and this unique_id value in itself doesn't provide any useful information for predicting chronic kidney disease as it has no relevance to patient's symptoms or lab results. Therefore, we should exclude this column from the dataset. 

In [11]:
#Remove that column
combined_data = combined_data.drop(columns='unique_id', axis=1)
# Print the dataset
print(combined_data)

          age    bp       bgr        bu        sc       sod       pot  \
2    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.310345   
149  0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.482759   
150  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.448276   
151  0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.206897   
152  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.620690   

         hemo       pcv      wbcc  ...  cad_no  cad_yes  appet_good  \
2    0.000000  0.032258  0.395161  ...       0      

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [12]:
# Export the dataframe to a new csv file
combined_data.to_csv('cleaned_data_results_csv.csv', index=False)
# Export the dataframe to a new json file
combined_data.to_json('cleaned_data_results_json.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [13]:
pd.set_option('display.max_columns', None)
print(combined_data.head(15))

         age    bp       bgr        bu        sc       sod       pot  \
2   0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5   0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7   0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8   0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9   0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
10  0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.034483   
12  0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.655172   
13  0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.344828   
15  0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
18  0.756757  0.25  0.223176  0.209150  0.160494  0.533333  0.620690   
19  0.662162  0.50  0.618026  0.411765  0.432099  0.566667  0.689655   
20  0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.862069   
22  0.878378  0.00  0.206009  0.751634  0.604938  0.533333  0.68

### Paste the markdown table here

## The markdown table is as follows:

|    | age     | bp    | bgr      | bu       | sc       | sod      | pot      | hemo     | pcv     | wbcc     | rbcc    | al   | su   | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|---------|-------|----------|----------|----------|----------|----------|----------|---------|----------|---------|------|------|--------------|------------|-------------|-----------|----------------|-------------|---------------|------------|--------|---------|-------|--------|--------|---------|------------|------------|-------|--------|--------|---------|------------|---------------|
| 2  | 0.743243| 0.50  | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.793103 | 0.000000 | 0.032258| 0.395161 | 0.057143| 2.0  | 0.0  | 1            | 0          | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 0      | 1       | 0          | 1          | 0     | 1      | 0      | 1       | 1          | 0             |
| 5  | 0.770270| 1.00  | 0.901288 | 0.163399 | 0.345679 | 0.766667 | 0.206897 | 0.524752 | 0.548387| 0.443548 | 0.342857| 2.0  | 2.0  | 0            | 1          | 0           | 1         | 1              | 0           | 0             | 1          | 0      | 1       | 1     | 0      | 0      | 1       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0             |
| 7  | 0.905405| 0.50  | 0.785408 | 0.862745 | 0.518519 | 0.600000 | 1.000000 | 0.277228 | 0.322581| 0.233871 | 0.371429| 2.0  | 0.0  | 1            | 0          | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 0      | 1       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0             |
| 8  | 0.202703| 0.75  | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.206897 | 0.059406 | 0.000000| 0.653226 | 0.257143| 4.0  | 0.0  | 0            | 1          | 1           | 0         | 0              | 1           | 0             | 1          | 1      | 0       | 1     | 0      | 1      | 0       | 1          | 0          | 1     | 0      | 0      | 1       | 1          | 0             |
| 9  | 0.783784| 1.00  | 0.399142 | 0.287582 | 0.839506 | 0.666667 | 0.586207 | 0.019802 | 0.096774| 0.258065 | 0.114286| 4.0  | 2.0  | 1            | 0          | 1           | 0         | 1              | 0           | 0             | 1          | 0      | 1       | 0     | 1      | 1      | 0       | 1          | 0          | 0     | 1      | 1      | 0       | 1          | 0             |
| 10 | 0.729730| 0.00  | 0.935622 | 0.169935 | 0.160494 | 0.333333 | 0.034483 | 0.019802 | 0.064516| 0.879032 | 0.000000| 3.0  | 1.0  | 0            | 1          | 1           | 0         | 0              | 1           | 1             | 0          | 0      | 1       | 1     | 0      | 1      | 0       | 0          | 1          | 1     | 0      | 0      | 1       | 1          | 0             |
| 12 | 0.675676| 0.75  | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.655172 | 0.138614 | 0.193548| 0.169355 | 0.114286| 2.0  | 0.0  | 1            | 0          | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 1     | 0      | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0             |
| 13 | 0.851351| 0.25  | 0.618026 | 0.562092 | 0.728395 | 0.000000 | 0.344828 | 0.168317 | 0.161290| 0.580645 | 0.085714| 4.0  | 3.0  | 0            | 1          | 1           | 0         | 0              | 1           | 0             | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 1          | 0          | 0     | 1      | 0      | 1       | 1          | 0             |
| 15 | 0.540541| 0.00  | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.379310 | 0.207921 | 0.161290| 0.830645 | 0.057143| 1.0  | 0.0  | 0            | 1          | 0           | 1         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0             |
| 18 | 0.756757| 0.25  | 0.223176 | 0.209150 | 0.160494 | 0.533333 | 0.620690 | 0.485149 | 0.516129| 0.290323 | 0.257143| 3.0  | 0.0  | 0            | 1          | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0             |
| 19 | 0.662162| 0.50  | 0.618026 | 0.411765 | 0.432099 | 0.566667 | 0.689655 | 0.316832 | 0.354839| 0.250000 | 0.200000| 3.0  | 1.0  | 0            | 1          | 1           | 0         | 0              | 1           | 0             | 1          | 0      | 1       | 0     | 1      | 1      | 0       | 1          | 0          | 0     | 1      | 1      | 0       | 1          | 0             |
| 20 | 0.783784| 0.00  | 0.725322 | 0.313725 | 0.481481 | 0.566667 | 0.862069 | 0.178218 | 0.193548| 0.258065 | 0.114286| 4.0  | 1.0  | 1            | 0          | 1           | 0         | 1              | 0           | 0             | 1          | 0      | 1       | 0     | 1      | 1      | 0       | 0          | 1          | 0     | 1      | 1      | 0       | 1          | 0             |
| 22 | 0.878378| 0.00  | 0.206009 | 0.751634 | 0.604938 | 0.533333 | 0.689655 | 0.366337 | 0.387097| 0.879032 | 0.371429| 4.0  | 0.0  | 0            | 1          | 0           | 1         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 0          | 1          | 0     | 1      | 1      | 0       | 1          | 0             |
| 23 | 0.878378| 0.25  | 0.639485 | 0.470588 | 0.395062 | 0.433333 | 0.517241 | 0.267327 | 0.322581| 0.104839 | 0.171429| 3.0  | 0.0  | 1            | 0          | 1           | 0         | 0              | 1           | 0             | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0             |
| 29 | 0.716216| 0.50  | 1.000000 | 0.163399 | 0.111111 | 0.066667 | 0.206897 | 0.267327 | 0.387097| 0.532258 | 0.371429| 1.0  | 0.0  | 1            | 0          | 0           | 1         | 1              | 0           | 1             | 0          | 1      | 0       | 0     | 1      | 1      | 0       | 0          | 1          | 1     | 0      | 1      | 0       | 1          | 0             |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

This can be done by always cross-checking the language model’s output against the original dataset to ensure accuracy. After performing a transformation or conversion suggested by a language model, we can verify the results programmatically by comparing summary statistics, row counts, and column names to ensure they match the expected values. This helps maintain data integrity and prevents errors during the transformation process.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [14]:
# Code to rename all the columns in the dataset
renaming_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rbcc': 'red blood cell count',
    'al': 'albumin',
    'su': 'sugar',
    'rbc_abnormal': 'red blood cells abnormal',
    'rbc_normal': 'red blood cells normal',
    'pc_abnormal': 'pus cell abnormal',
    'pc_normal': 'pus cell normal',
    'pcc_notpresent': 'pus cell clumps not present',
    'pcc_present': 'pus cell clumps present',
    'ba_notpresent': 'bacteria not present',
    'ba_present': 'bacteria present',
    'htn_no': 'hypertension no',
    'htn_yes': 'hypertension yes',
    'dm_no': 'diabetes mellitus no',
    'dm_yes': 'diabetes mellitus yes',
    'cad_no': 'coronary artery disease no',
    'cad_yes': 'coronary artery disease yes',
    'appet_good': 'appetite good',
    'appet_poor': 'appetite poor',
    'pe_no': 'pedal edema no',
    'pe_yes': 'pedal edema yes',
    'ane_no': 'anemia no',
    'ane_yes': 'anemia yes',
    'Target_ckd': 'class ckd',
    'Target_notckd': 'class not ckd'
}
combnined_data = combined_data.rename(columns=renaming_mapping, inplace=True)
print(combined_data)

          age  blood pressure  blood glucose random  blood urea  \
2    0.743243            0.50              0.442060    0.901961   
5    0.770270            1.00              0.901288    0.163399   
7    0.905405            0.50              0.785408    0.862745   
8    0.202703            0.75              0.158798    0.196078   
9    0.783784            1.00              0.399142    0.287582   
..        ...             ...                   ...         ...   
148  0.189189            0.25              0.227468    0.222222   
149  0.216216            0.00              0.115880    0.052288   
150  0.513514            0.00              0.107296    0.235294   
151  0.702703            0.50              0.128755    0.261438   
152  0.689189            0.00              0.150215    0.254902   

     serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
2            0.432099  0.500000   0.793103    0.000000            0.032258   
5            0.345679  0.766667   0.206

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [15]:
# Converted SQL to Pandas code
# for the sake of the question, I am assuming that the 'Target' column is the class column because after I did the encoding of the target column, the column name changed to 'Target_notckd' and 'Target_ckd'.
# here is how it would look like if I did it for the 'Target_notckd' column and 'Target_ckd' column
target_count1 = combined_data.groupby('class ckd').size().reset_index(name='count')
target_count2 = combined_data.groupby('class not ckd').size().reset_index(name='count')
# here is how I would do it if I did it for the 'Target' column like the SQL query states:
# target_count = combined_data.groupby('Target').size().reset_index(name='count')
print('Class CKD\n')
print(target_count1)
print()
print('Class Not CKD\n')
print(target_count2)

Class CKD

   class ckd  count
0          0    111
1          1     22

Class Not CKD

   class not ckd  count
0              0     22
1              1    111


*Describe what the above code does here*

The following dexcribes what the code above does, which is the output of the `class ckd` and `class notckd` columns of the DataFrame:

The code counts the number of occurrences for each unique value in the columns `class ckd` and `class not ckd` and stores these counts in separate DataFrames, `target_count1` and `target_count2`. It then prints these counts to display how many records belong to each class (0 and 1) for `class ckd` and `class not ckd`.

The following describes what the code would do had my code have done it for a `Target` column in a DataFrame:

This code groups the `combined_data` DataFrame by the values in the `Target` column and calculates the number of rows for each unique value. It then creates a new DataFrame called `target_count` that contains two columns: `Target` and `count`. The `Target` column represents each unique value in the original `Target` column, while the `count` column shows how many times each of those values appears in the dataset.

Overall, it shows us how many positive and negative cases are present in the dataset based on the `Target` columns or in my case, the `class ckd` and `class not ckd` columns.